GitUI/BranchTreePanel/RepoObjectsTree.Nodes.cs

```diff
@@ -117,43 +151,55 @@ protected Tree(TreeNode treeNode, IGitUICommandsSource uiCommands)
             public bool IgnoreSelectionChangedEvent { get; set; }
             protected GitModule Module => UICommands.Module;
 
-            public async Task ReloadAsync(CancellationToken token)
+            // Invoke from child class to reload nodes for the current Tree. Clears Nodes, invokes
+            // input async function that should populate Nodes, then fills the tree view with its contents,
+            // making sure to disable/enable the control.
+            protected void ReloadNodes(Func<CancellationToken, Task> loadNodesTask)
             {
-                await TreeViewNode.TreeView.SwitchToMainThreadAsync(token);
-                ClearNodes();
-
-                await LoadNodesAsync(token).ConfigureAwait(false);
+                ThreadHelper.ThrowIfNotOnUIThread();
+                TreeViewNode.TreeView.Enabled = false;
+                Nodes.Clear();
 
-                await TreeViewNode.TreeView.SwitchToMainThreadAsync(token);
-                TreeViewNode.TreeView.BeginUpdate();
-                try
+                ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
                 {
-                    FillTreeViewNode();
-                    if (TreeViewNode.TreeView.SelectedNode != null)
+                    try
                     {
-                        TreeViewNode.TreeView.SelectedNode.EnsureVisible();
+                        var token = _reloadCancellation.Next();
+                        await loadNodesTask(token).ConfigureAwait(false);
+
+                        token.ThrowIfCancellationRequested();
+                        await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(token);
```

Comment:
Yeah, so what you said is what I thought, and my original push didn't have the ConfigureAwait(false). But then @gerhardol said he had a 30s stall when refreshing, and when I tested, I had some 5s stalls here and there. Comparing against the original code, I found that the await LoadNodesAsync would ConfigureAwait(false), despite joining on the UI thread right after. I tried it, and indeed it seems to no longer stall for that long.

I was wondering if perhaps ConfigureAwait(false) is also used as a hint to the task manager to run the task on a non-UI thread; or conversely, not having the ConfigureAwait(false) may make it more likely to run on the UI thread, which might cause thread contention, and thus bad performance? I don't know, I'm fairly new to async-await in C# (as as aside, I think it's really cool).

Reply:
Yes `ConfigureAwait(false)` tells the awaiter to not capture the current `SynchronizationContext` (which in this case would be `WinFormsSynchronizationContext`) which has the effect of scheduling the continuation on the thread pool rather than the UI thread.

Removing `ConfigureAwait` it can highlight other threading issues. Possibly `loadNodesTask` is trying to do something on the UI thread. If you pause into the debugger when you have one of these slowdowns and look at the "parallel stacks" view of Visual Studio, you'll likely see what's going on.